File: src/resource-orchestrator.js
/**
 * ROBBIE RESOURCE ORCHESTRATOR
 * Intelligent network-aware resource discovery and routing
 * 
 * Priority Order:
 * 1. Aurora Town (always first choice - has RunPod GPU access)
 * 2. Vengeance (local RTX 4090)
 * 3. RobbieBook1 (M3 MacBook Pro - local fallback)
 * 
 * Features:
 * - Automatic resource discovery
 * - Network-aware routing
 * - Graceful degradation (airplane mode = local)
 * - Database sync/fallback
 */

class ResourceOrchestrator {
    constructor() {
        this.resources = {
            aurora: {
                name: 'Aurora Town',
                priority: 1,
                endpoints: {
                    chat: 'http://10.0.0.1:8007/api/chat',
                    llm: 'http://10.0.0.1:11434',
                    db: 'postgresql://aurora-town-u44170.vm.elestio.app:5432/aurora'
                },
                capabilities: ['chat', 'llm', 'database', 'runpod-gpu', 'vector-search'],
                status: 'unknown',
                lastCheck: null,
                latency: null
            },
            vengeance: {
                name: 'Vengeance',
                priority: 2,
                endpoints: {
                    chat: 'http://localhost:3000/api/chat',
                    llm: 'http://localhost:11434',
                    db: 'postgresql://localhost:5432/vengeance',
                    gpu: 'http://localhost:8001/gpu'
                },
                capabilities: ['chat', 'llm', 'database', 'rtx-4090', 'local-training'],
                status: 'unknown',
                lastCheck: null,
                latency: null
            },
            robbiebook1: {
                name: 'RobbieBook1',
                priority: 3,
                endpoints: {
                    chat: 'http://localhost:8080/api/chat',
                    llm: 'http://localhost:11434',
                    db: 'sqlite:///Users/allan/robbie-local.db'
                },
                capabilities: ['chat', 'llm', 'local-db', 'm3-neural-engine'],
                status: 'unknown',
                lastCheck: null,
                latency: null
            }
        };

        this.currentResource = null;
        this.offlineMode = false;
        this.discoveryInterval = null;
    }

    /**
     * Initialize orchestrator - discover available resources
     */
    async initialize() {
        console.log('🚀 Robbie Resource Orchestrator starting...');
        
        // Immediate discovery
        await this.discoverResources();
        
        // Continuous monitoring (every 30 seconds)
        this.discoveryInterval = setInterval(() => {
            this.discoverResources();
        }, 30000);

        // Listen for network changes
        if (typeof window !== 'undefined') {
            window.addEventListener('online', () => this.handleNetworkChange(true));
            window.addEventListener('offline', () => this.handleNetworkChange(false));
        }

        console.log(`✅ Orchestrator initialized. Current resource: ${this.currentResource?.name || 'None'}`);
        return this.currentResource;
    }

    /**
     * Discover and test all available resources
     */
    async discoverResources() {
        console.log('🔍 Discovering available resources...');
        
        const discoveries = Object.entries(this.resources).map(async ([key, resource]) => {
            try {
                const startTime = Date.now();
                
                // Try to ping the chat endpoint
                const controller = new AbortController();
                const timeoutId = setTimeout(() => controller.abort(), 2000); // 2s timeout
                
                const response = await fetch(resource.endpoints.chat.replace('/api/chat', '/health'), {
                    method: 'GET',
                    signal: controller.signal
                });
                
                clearTimeout(timeoutId);
                
                if (response.ok) {
                    resource.status = 'online';
                    resource.latency = Date.now() - startTime;
                    resource.lastCheck = new Date();
                    console.log(`✅ ${resource.name}: ONLINE (${resource.latency}ms)`);
                } else {
                    resource.status = 'degraded';
                    console.log(`⚠️  ${resource.name}: DEGRADED (HTTP ${response.status})`);
                }
            } catch (error) {
                resource.status = 'offline';
                resource.latency = null;
                console.log(`❌ ${resource.name}: OFFLINE (${error.message})`);
            }
        });

        await Promise.all(discoveries);

        // Select best available resource
        this.selectBestResource();
    }

    /**
     * Select the best available resource based on priority and status
     */
    selectBestResource() {
        // Sort by priority, then by status
        const available = Object.values(this.resources)
            .filter(r => r.status === 'online')
            .sort((a, b) => a.priority - b.priority);

        if (available.length > 0) {
            this.currentResource = available[0];
            this.offlineMode = false;
            console.log(`🎯 Selected: ${this.currentResource.name} (Priority ${this.currentResource.priority}, ${this.currentResource.latency}ms)`);
        } else {
            // No network resources - go local
            this.currentResource = this.resources.robbiebook1;
            this.offlineMode = true;
            console.log('✈️  OFFLINE MODE: Using local RobbieBook1 resources');
        }

        // Emit event for UI updates
        this.emitResourceChange();
    }

    /**
     * Handle network connectivity changes
     */
    async handleNetworkChange(isOnline) {
        if (isOnline) {
            console.log('🌐 Network restored - rediscovering resources...');
            await this.discoverResources();
        } else {
            console.log('✈️  Network lost - switching to offline mode...');
            this.currentResource = this.resources.robbiebook1;
            this.offlineMode = true;
            this.emitResourceChange();
        }
    }

    /**
     * Send a chat message using best available resource
     */
    async sendMessage(message, options = {}) {
        if (!this.currentResource) {
            throw new Error('No resources available');
        }

        console.log(`💬 Sending message via ${this.currentResource.name}...`);

        try {
            const response = await fetch(this.currentResource.endpoints.chat, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    message,
                    resource: this.currentResource.name,
                    offline_mode: this.offlineMode,
                    ...options
                })
            });

            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }

            return response;
        } catch (error) {
            console.error(`❌ Failed to send via ${this.currentResource.name}:`, error);
            
            // Try to failover to next available resource
            return this.failoverAndRetry(message, options);
        }
    }

    /**
     * Failover to next available resource and retry
     */
    async failoverAndRetry(message, options) {
        console.log('🔄 Attempting failover...');
        
        // Mark current resource as degraded
        if (this.currentResource) {
            this.currentResource.status = 'degraded';
        }

        // Rediscover and select new resource
        await this.discoverResources();

        if (!this.currentResource || this.currentResource.status !== 'online') {
            throw new Error('All resources unavailable');
        }

        // Retry with new resource
        return this.sendMessage(message, options);
    }

    /**
     * Get current resource info for UI display
     */
    getResourceInfo() {
        if (!this.currentResource) {
            return {
                name: 'No Resource',
                status: 'offline',
                mode: 'offline',
                hardware: 'Unknown',
                capabilities: []
            };
        }

        return {
            name: this.currentResource.name,
            status: this.currentResource.status,
            mode: this.offlineMode ? 'offline' : 'online',
            latency: this.currentResource.latency,
            hardware: this.getHardwareInfo(this.currentResource),
            capabilities: this.currentResource.capabilities,
            lastCheck: this.currentResource.lastCheck
        };
    }

    /**
     * Get hardware info string for display
     */
    getHardwareInfo(resource) {
        if (resource === this.resources.aurora) {
            return 'CPU: AMD EPYC (+ RunPod GPU)';
        } else if (resource === this.resources.vengeance) {
            return 'GPU: RTX 4090';
        } else if (resource === this.resources.robbiebook1) {
            return 'CPU: M3 Pro';
        }
        return 'Unknown';
    }

    /**
     * Emit resource change event for UI updates
     */
    emitResourceChange() {
        if (typeof window !== 'undefined') {
            const event = new CustomEvent('resourcechange', {
                detail: this.getResourceInfo()
            });
            window.dispatchEvent(event);
        }
    }

    /**
     * Get database connection string for current resource
     */
    getDatabaseEndpoint() {
        if (!this.currentResource) {
            return null;
        }
        return this.currentResource.endpoints.db;
    }

    /**
     * Get LLM endpoint for current resource
     */
    getLLMEndpoint() {
        if (!this.currentResource) {
            return null;
        }
        return this.currentResource.endpoints.llm;
    }

    /**
     * Check if a specific capability is available
     */
    hasCapability(capability) {
        if (!this.currentResource) {
            return false;
        }
        return this.currentResource.capabilities.includes(capability);
    }

    /**
     * Cleanup
     */
    destroy() {
        if (this.discoveryInterval) {
            clearInterval(this.discoveryInterval);
        }
    }
}

// Export for use in Node.js and browser
if (typeof module !== 'undefined' && module.exports) {
    module.exports = ResourceOrchestrator;
}

if (typeof window !== 'undefined') {
    window.ResourceOrchestrator = ResourceOrchestrator;
}

File: apps/archive-legacy/backend/app/services/resource_orchestrator.py
"""
ROBBIE RESOURCE ORCHESTRATOR
Intelligent network-aware resource discovery and routing over SSH mesh

Priority Order:
1. Vengeance RTX 4090 (via reverse SSH tunnel on localhost:11434)
2. Local Aurora CPU (localhost:11434 direct)
3. RunPod GPU (when available)
"""

import asyncio
import aiohttp
import logging
from typing import Optional, Dict, List
from datetime import datetime
import psutil

logger = logging.getLogger(__name__)


class ResourceNode:
    def __init__(self, name: str, priority: int, endpoints: Dict[str, str], capabilities: List[str]):
        self.name = name
        self.priority = priority
        self.endpoints = endpoints
        self.capabilities = capabilities
        self.status = 'unknown'
        self.last_check = None
        self.latency = None
        self.gpu_available = False
        self.gpu_name = None


class ResourceOrchestrator:
    def __init__(self):
        self.nodes = {
            'vengeance': ResourceNode(
                name='Vengeance RTX 4090',
                priority=1,
                endpoints={
                    'ollama': 'http://localhost:11434',  # Via reverse SSH tunnel
                    'gpu': 'http://localhost:8001',
                    'db': 'postgresql://localhost:5433/vengeance'
                },
                capabilities=['chat', 'llm', 'rtx-4090', 'gpu-inference']
            ),
            'aurora': ResourceNode(
                name='Aurora Town CPU',
                priority=2,
                endpoints={
                    'ollama': 'http://localhost:11434',  # Direct local
                    'db': 'postgresql://localhost:5432/aurora'
                },
                capabilities=['chat', 'llm', 'cpu-inference', 'database']
            ),
            'runpod': ResourceNode(
                name='RunPod GPU',
                priority=3,
                endpoints={
                    'ollama': 'http://localhost:11435',  # Via SSH tunnel (when configured)
                },
                capabilities=['chat', 'llm', 'gpu-inference', 'high-vram']
            )
        }
        
        self.current_node = None
        self.discovery_task = None
        self._session = None

    async def initialize(self):
        """Initialize orchestrator and start resource discovery"""
        logger.info("🚀 Robbie Resource Orchestrator starting...")
        
        # Create aiohttp session
        self._session = aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=2))
        
        # Initial discovery
        await self.discover_resources()
        
        # Start continuous monitoring (every 30 seconds)
        self.discovery_task = asyncio.create_task(self._continuous_discovery())
        
        logger.info(f"✅ Orchestrator initialized. Current node: {self.current_node.name if self.current_node else 'None'}")
        return self.current_node

    async def _continuous_discovery(self):
        """Background task for continuous resource discovery"""
        while True:
            try:
                await asyncio.sleep(30)
                await self.discover_resources()
            except Exception as e:
                logger.error(f"Error in continuous discovery: {e}")

    async def discover_resources(self):
        """Discover and test all available resources"""
        logger.info("🔍 Discovering available resources...")
        
        # Test all nodes in parallel
        tasks = []
        for node_name, node in self.nodes.items():
            tasks.append(self._test_node(node_name, node))
        
        await asyncio.gather(*tasks, return_exceptions=True)
        
        # Select best available resource
        self.select_best_resource()

    async def _test_node(self, node_name: str, node: ResourceNode):
        """Test if a node is available and responsive"""
        try:
            start_time = asyncio.get_event_loop().time()
            
            # Try to ping the Ollama endpoint
            async with self._session.get(f"{node.endpoints['ollama']}/api/tags") as response:
                if response.status == 200:
                    data = await response.json()
                    node.status = 'online'
                    node.latency = int((asyncio.get_event_loop().time() - start_time) * 1000)
                    node.last_check = datetime.now()
                    
                    # Check if it's a GPU node by looking at available models
                    if 'models' in data and len(data['models']) > 0:
                        # Vengeance has GPU if we can reach it on the tunnel
                        if node_name == 'vengeance':
                            node.gpu_available = True
                            node.gpu_name = 'RTX 4090'
                    
                    logger.info(f"✅ {node.name}: ONLINE ({node.latency}ms)")
                else:
                    node.status = 'degraded'
                    logger.warning(f"⚠️  {node.name}: DEGRADED (HTTP {response.status})")
        except asyncio.TimeoutError:
            node.status = 'offline'
            node.latency = None
            logger.debug(f"❌ {node.name}: TIMEOUT")
        except Exception as e:
            node.status = 'offline'
            node.latency = None
            logger.debug(f"❌ {node.name}: OFFLINE ({str(e)})")

    def select_best_resource(self):
        """Select the best available resource based on priority and status"""
        # Get all online nodes sorted by priority
        available = [
            node for node in self.nodes.values()
            if node.status == 'online'
        ]
        available.sort(key=lambda x: x.priority)
        
        if available:
            self.current_node = available[0]
            logger.info(f"🎯 Selected: {self.current_node.name} (Priority {self.current_node.priority}, {self.current_node.latency}ms)")
        else:
            # Fallback to Aurora (always assume local is available)
            self.current_node = self.nodes['aurora']
            logger.warning("⚠️  No nodes responding, using Aurora CPU fallback")

    def get_resource_info(self) -> Dict:
        """Get current resource info for UI display"""
        if not self.current_node:
            return {
                'name': 'No Resource',
                'status': 'offline',
                'hardware': 'Unknown',
                'capabilities': []
            }
        
        return {
            'name': self.current_node.name,
            'status': self.current_node.status,
            'latency': self.current_node.latency,
            'hardware': self._get_hardware_info(self.current_node),
            'capabilities': self.current_node.capabilities,
            'last_check': self.current_node.last_check.isoformat() if self.current_node.last_check else None,
            'gpu_available': self.current_node.gpu_available,
            'gpu_name': self.current_node.gpu_name
        }

    def _get_hardware_info(self, node: ResourceNode) -> str:
        """Get hardware info string for display"""
        if node.gpu_available and node.gpu_name:
            return f"GPU: {node.gpu_name}"
        elif 'vengeance' in node.name.lower():
            return 'GPU: RTX 4090'
        elif 'aurora' in node.name.lower():
            return 'CPU: AMD EPYC'
        elif 'runpod' in node.name.lower():
            return 'GPU: RunPod'
        return 'Unknown'

    def get_ollama_endpoint(self) -> str:
        """Get the Ollama endpoint for the current resource"""
        if not self.current_node:
            return 'http://localhost:11434'
        return self.current_node.endpoints.get('ollama', 'http://localhost:11434')

    def get_database_endpoint(self) -> str:
        """Get the database endpoint for the current resource"""
        if not self.current_node:
            return None
        return self.current_node.endpoints.get('db')

    def has_capability(self, capability: str) -> bool:
        """Check if current resource has a specific capability"""
        if not self.current_node:
            return False
        return capability in self.current_node.capabilities

    async def failover_and_retry(self):
        """Mark current node as degraded and select next best resource"""
        if self.current_node:
            logger.warning(f"🔄 Failing over from {self.current_node.name}")
            self.current_node.status = 'degraded'
        
        await self.discover_resources()
        
        if not self.current_node or self.current_node.status != 'online':
            raise Exception("All resources unavailable")

    async def cleanup(self):
        """Cleanup resources"""
        if self.discovery_task:
            self.discovery_task.cancel()
            try:
                await self.discovery_task
            except asyncio.CancelledError:
                pass
        
        if self._session:
            await self._session.close()


# Global orchestrator instance
_orchestrator: Optional[ResourceOrchestrator] = None


async def get_orchestrator() -> ResourceOrchestrator:
    """Get or create the global orchestrator instance"""
    global _orchestrator
    if _orchestrator is None:
        _orchestrator = ResourceOrchestrator()
        await _orchestrator.initialize()
    return _orchestrator


async def get_current_resource_info() -> Dict:
    """Get current resource info (convenience function)"""
    orchestrator = await get_orchestrator()
    return orchestrator.get_resource_info()
